dask/async.py

```diff
@@ -458,16 +463,28 @@ def fire_task():
             data = dict((dep, state['cache'][dep])
                         for dep in get_dependencies(dsk, key))
             # Submit
-            apply_async(execute_task, args=[key, dsk[key], data, queue,
-                                            get_id, raise_on_exception])
+            apply_async(execute_task,
+                        args=(key, marshall((dsk[key], data)),
+                              marshall, unmarshall,
+                              get_id, raise_on_exception),
+                        callback=queue.put)
+
+        def fill_pool_with_ready_tasks():
+            while state['ready'] and len(state['running']) < num_workers:
+                fire_task()
 
         # Seed initial tasks into the thread pool
-        while state['ready'] and len(state['running']) < num_workers:
-            fire_task()
+        fill_pool_with_ready_tasks()
 
         # Main loop, wait on tasks to finish, insert new ones
         while state['waiting'] or state['ready'] or state['running']:
-            key, res, tb, worker_id = queue.get()
+            key, res_info = queue.get()
+            try:
+                res, tb, worker_id = unmarshall(res_info)
+            except Exception as e:
```

Comment:
What I was thinking of is in `execute_task`, where we take care when putting even exceptions into the queue so that, if the placing of the exception fails we are still ok.  

My apologies for not reading this more thoroughly before commenting earlier.